src/wavm/WAVMWasmModule.cpp

```diff
@@ -1003,86 +1010,176 @@ U32 WAVMWasmModule::mmapFile(U32 fd, U32 length)
     return wasmPtr;
 }
 
-U32 WAVMWasmModule::allocateThreadStack()
+U32 WAVMWasmModule::growMemory(U32 nBytes)
 {
-    createMemoryGuardRegion();
-    U32 wasmPtr = mmapMemory(THREAD_STACK_SIZE);
-    createMemoryGuardRegion();
-
-    return wasmPtr;
-}
+    auto logger = faabric::util::getLogger();
 
-U32 WAVMWasmModule::mmapMemory(U32 length)
-{
-    // Round up to page boundary
-    Uptr nWasmPages = getNumberOfWasmPagesForBytes(length);
-    return mmapPages(nWasmPages);
-}
+    if (nBytes == 0) {
+        return currentBrk;
+    }
 
-U32 WAVMWasmModule::mmapPages(U32 pages)
-{
     // Cautious locking
     faabric::util::UniqueLock lock(moduleMemoryMutex);
 
-    const std::shared_ptr<spdlog::logger>& logger = faabric::util::getLogger();
-    U64 maxSize = getMemoryType(defaultMemory).size.max;
-    Uptr currentPageCount = Runtime::getMemoryNumPages(defaultMemory);
+    size_t oldBytes = getMemorySizeBytes();
+    size_t newBytes = oldBytes + nBytes;
+
+    uint32_t oldBrk = currentBrk;
+    uint32_t newBrk = currentBrk + nBytes;
 
-    if (pages == 0) {
-        throw std::runtime_error("Requesting mapping of zero pages");
+    if (!isWasmPageAligned(newBrk)) {
+        logger->error("Growing memory by {} is not wasm page aligned", nBytes);
+        throw std::runtime_error("Non-wasm-page-aligned memory growth");
```

Comment:
We check wasm page alignment in all these functions because they should only ever be called behind the scenes by libc and malloc. Where we call them internally, we have to make sure we round up to the nearest wasm page.